### PR TITLE
Add cache

### PR DESCRIPTION
**Pull request by adriantpaez**
## Changes:
- Add the `internal/cache` package with the `Cache` interface and the `LRUCache` implementation

## Types of changes

What types of changes does your code introduce?

- New feature (non-breaking change which adds functionality)

## Testing
**Requires testing**

- Yes

**In case you checked yes, did you write tests??**

- Yes

### CODE REVIEW

**Review by stdevMac — APPROVED** [review]

File: internal/cache/key.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package cache
+
+import "crypto/sha1"
+
+type chacheKey [sha1.Size]byte
+
+func newCacheKey(v []byte) chacheKey {
+	return sha1.Sum(v)
```
Comment (tshakalekholoane) [inline internal/cache/key.go:8]: Is it important that we use a cryptographic hash? It seems to me the only property we care about here is that the key is unique so we might be better off with something less costly such as the [maphash](https://pkg.go.dev/hash/maphash) or [fnv](https://pkg.go.dev/hash/fnv) from the standard library.

**Review by tshakalekholoane — COMMENTED** [review]
To keep consistent with Go conventions maybe instead of `Count` we could do `Len` (see [`bytes.Buffer.Len`](https://pkg.go.dev/bytes#Buffer.Len)) and similarly with the `Capacity` (see [`bytes.Buffer.Cap`](https://pkg.go.dev/bytes#Buffer.Cap)). Otherwise this looks great!

File: internal/cache/key.go (hunk 1 above)
Comment (adriantpaez) [reply internal/cache/key.go:8]: I will change it to maphash

**Comment by IronGauntlets** [comment]
Since this is a new package can we please move away from the convention of using internal packages? I would suggest adding this package under `juno/cache`.

File: internal/cache/lru.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// LRUCache is a cache with the least-recently-used policy.
+type LRUCache struct {
+	hashMap  map[chacheKey]*cacheNode
+	start    *cacheNode
+	end      *cacheNode
+	count    int
+	capacity int
+	lock     sync.Mutex
```
Comment (IronGauntlets) [inline internal/cache/lru.go:21]: Is the cache being used somewhere that is accessed by multiple go routine? 

If not, I would recommend not adding a mutex at this stage.

File: internal/cache/lru.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+// the least-recently-used key-value pair is removed.
+func (c *LRUCache) Put(k []byte, v []byte) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// Build the cache key and node
+	key := newCacheKey(k)
+	node := &cacheNode{Key: key, Value: v}
+
+	// Cache is empty
```
Comment (IronGauntlets) [inline internal/cache/lru.go:42]: Probably unnecessary. Otherwise, please expand with what happens when cache is empty

**Comment by adriantpaez** [comment]
> Since this is a new package can we please move away from the convention of using internal packages? I would suggest adding this package under `juno/cache`. Where the cache interface is stored under file `cache.go` and the lru implementation is stored under `juno/cache/lru` directory.

@IronGauntlets Sounds good to me, but because now the project structure is under the `internal` and `pkg` directories, we need to agree first if we will continue with that structure or move to the project root because I can't break the current structure

File: internal/cache/lru.go (hunk 2 above)
Comment (adriantpaez) [reply internal/cache/lru.go:23]: I think cache will be used for multiple RPC concurrently threads with certain managers

Comment (jmederosalvarado) [reply internal/cache/lru.go:23]: > Is the cache being used somewhere that is accessed by multiple go routine?
> If not, I would recommend not adding a mutex at this stage.

The main purpose of the cache is to have it as a layer between the database and initially the trie, but probably all database access will use a cache, having it not thread-safe would be the same as having the db not thread-safe.

**Review by jmederosalvarado — CHANGES_REQUESTED** [review]

**Pull request merged**